src/twisted/web/_websocket_impl.py

```diff
@@ -0,0 +1,457 @@
+# -*- test-case-name: twisted.web.test.test_websocket -*-
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from functools import singledispatch
+from typing import Callable, Generic, Protocol as TypingProtocol, TypeVar, Union
+
+from zope.interface import implementer
+
+from hyperlink import URL
+from wsproto import Connection, ConnectionType, WSConnection
+from wsproto.connection import ConnectionState
+from wsproto.events import (
+    AcceptConnection,
+    BytesMessage,
+    CloseConnection,
+    Event,
+    Ping,
+    Pong,
+    RejectConnection,
+    RejectData,
+    Request as WSRequest,
+    TextMessage,
+)
+from wsproto.handshake import H11Handshake
+from wsproto.utilities import RemoteProtocolError
+
+from twisted.internet.defer import Deferred
+from twisted.internet.interfaces import (
+    IConsumer,
+    IProtocol,
+    IPushProducer,
+    IReactorTCP,
+    ITransport,
+)
+from twisted.internet.protocol import Factory as ProtocolFactory
+from twisted.logger import Logger
+from twisted.python.failure import Failure
+from twisted.web._responses import BAD_REQUEST
+from twisted.web.client import (
+    URI,
+    BrowserLikePolicyForHTTPS,
+    Response,
+    _StandardEndpointFactory,
+)
+from twisted.web.http_headers import Headers
+from twisted.web.iweb import IAgentEndpointFactory, IPolicyForHTTPS
+from twisted.web.resource import Resource
+from twisted.web.server import NOT_DONE_YET, Request
+
+
+class WebSocketTransport(TypingProtocol):
+    """
+    The transport that can send websocket messages.
+    """
+
+    def sendTextMessage(self, text: str) -> None:
+        """
+        Send a text message.
+        """
+
+    def sendBytesMessage(self, data: bytes) -> None:
+        """
+        Send a bytes message.
+        """
+
+    def loseConnection(self, code: int = 1000) -> None:
+        """
+        Drop the websocket connection.
+        """
+
+    def ping(self, payload: bytes = b"") -> None:
```

Comment:
Do we want to expose this API?

If yes, I think that we should add this to the example code.

Reply:
I added more details in `pongReceived` which will hopefully make it clearer what this is for.